File: day13_part2_2023.js
const fs = require('fs');

function parseInput(input) {
    const mirrors = [];

    let mirrorStr = [];
    input.forEach(line => {
        if (line === "") {
            mirrors.push(parseMirror(mirrorStr));
            mirrorStr = [];
        } else {
            mirrorStr.push(line);
        }
    });
    mirrors.push(parseMirror(mirrorStr));

    return mirrors;
}

function parseMirror(mirrorStr) {
    const mirror = {
        Rows: new Array(mirrorStr.length).fill(0),
        Cols: new Array(mirrorStr[0].length).fill(0),
    };

    mirrorStr.forEach((line, y) => {
        line.split('').forEach((char, x) => {
            mirror.Rows[y] <<= 1;
            mirror.Cols[x] <<= 1;
            if (char === '#') {
                mirror.Rows[y]++;
                mirror.Cols[x]++;
            }
        });
    });

    return mirror;
}

function getMirrorAxis(lines) {
    for (let i = 1; i < lines.length; i++) {
        let isMirror = true;

        for (let j = 0; isMirror && j < Math.min(i, lines.length - i); j++) {
            if (lines[i - 1 - j] !== lines[i + j]) {
                isMirror = false;
            }
        }

        if (isMirror) {
            return i;
        }
    }

    return 0;
}

function getMirrorAxisWithOneSmudge(lines) {
    for (let i = 1; i < lines.length; i++) {
        let isMirror = true;
        let numSmudges = 0;

        for (let j = 0; isMirror && j < Math.min(i, lines.length - i); j++) {
            if (lines[i - 1 - j] !== lines[i + j]) {
                if (numSmudges > 0) {
                    isMirror = false;
                } else {
                    const dif = lines[i - 1 - j] ^ lines[i + j];
                    const isOnlyOneSmudge = (dif & (dif - 1)) === 0;
                    if (isOnlyOneSmudge) {
                        numSmudges++;
                    } else {
                        isMirror = false;
                    }
                }
            }
        }

        if (isMirror && numSmudges === 1) {
            return i;
        }
    }

    return 0;
}

function solve(input) {
    const mirrors = parseInput(input);

    let res = 0;
    mirrors.forEach(mirror => {
        res += getMirrorAxisWithOneSmudge(mirror.Cols);
        res += getMirrorAxisWithOneSmudge(mirror.Rows) * 100;
    });
    return res;
}

function readFile(fileName) {
    const file = fs.readFileSync(fileName, 'utf8');
    return file.trim().split('\n');
}

const input = readFile("input.txt");
console.log(solve(input));

File: haskell/day13_part2_2023.hs

import Data.Bits
import Data.List

parseMirror :: [String] -> ([Int], [Int])
parseMirror mirrorStr = (rows, cols)
  where
    rows = map rowToInt mirrorStr
    cols = map colToInt $ transpose mirrorStr
    rowToInt row = foldl' (\acc c -> (acc `shiftL` 1) .|. if c == '#' then 1 else 0) 0 row
    colToInt col = foldl' (\acc c -> (acc `shiftL` 1) .|. if c == '#' then 1 else 0) 0 col

getMirrorAxis :: [Int] -> Int
getMirrorAxis lines = head $ [i | i <- [1..length lines - 1], isMirror i] ++ [0]
  where
    isMirror i = all (\j -> lines !! (i-1-j) == lines !! (i+j)) [0..min (i-1) (length lines - i -1)]

getMirrorAxisWithSmudge :: [Int] -> Int
getMirrorAxisWithSmudge lines = head $ [i | i <- [1..length lines - 1], isMirror i] ++ [0]
  where
    isMirror i = (== 1) . sum $ [countBits (lines !! (i-1-j) `xor` lines !! (i+j)) | j <- [0..min (i-1) (length lines - i -1)]]
    countBits 0 = 0
    countBits x = (x .&. 1) + countBits (x `shiftR` 1)

solve :: [String] -> Int
solve inputLines = sum [getMirrorAxisWithSmudge cols + 100 * getMirrorAxisWithSmudge rows | (rows, cols) <- mirrors]
  where
    mirrors = map parseMirror (splitMirrors inputLines)
    splitMirrors [] = []
    splitMirrors lines = firstMirror : splitMirrors (drop 1 restLines)
      where
        (firstMirror, restLines) = break null lines

main :: IO ()
main = do
  inputLines <- readFile "input.txt"
  print $ solve (lines (init inputLines))
